src/stash.c

```diff
@@ -398,28 +398,23 @@ static int commit_worktree(
 	git_commit *b_commit,
 	git_commit *u_commit)
 {
-	int error = 0;
-	git_tree *w_tree = NULL, *i_tree = NULL;
-	git_index *i_index = NULL;
 	const git_commit *parents[] = {	NULL, NULL,	NULL };
-	int ignorecase;
+	git_index *i_index = NULL, *r_index = NULL;
+	git_tree *w_tree = NULL;
+	int error = 0, ignorecase;
 
 	parents[0] = b_commit;
 	parents[1] = i_commit;
 	parents[2] = u_commit;
 
-	if ((error = git_commit_tree(&i_tree, i_commit)) < 0)
-		goto cleanup;
-
-	if ((error = git_index_new(&i_index)) < 0 ||
-		(error = git_repository__cvar(&ignorecase, repo, GIT_CVAR_IGNORECASE)) < 0)
-		goto cleanup;
+	if ((error = git_repository_index(&r_index, repo) < 0) ||
+	    (error = git_index_new(&i_index)) < 0 ||
+	    (error = git_index__fill(i_index, &r_index->entries) < 0) ||
```

Comment:
Hmm, I haven't done a deep analysis here, but is there not a behavior difference?  Previously, we'd create an index from the commit tree, now we're using the existing index.  So if there are staged changes, those will now be included in `i_index` while they were not before.

(Apologies if there's already a guard here to avoid that).

I think that we have a stat-preserving way to make these match if that's needed.  I'm reasonably sure that `git_index_read_tree` will actually preserve the stat cache.  So adding that back after filling `i_index` with stat information should keep `i_index` identical to `i_tree` with the correct stat information (for matching entries).

But maybe I'm off base here and this function only gets called when `i_index` == HEAD, in which case, there's no reason to worry about that.  Indeed, I would expect unit tests to catch this case, but I wanted to mention it while it was top of mind for me.

I can do a more detailed 👀 soon.

Reply:
Any chance you're free to double check this soon :smile: @ethomson.